docs: Add a doc for devworkspace backup feature (#3028)
codenvy/che-docs: modules/administration-guide/pages/devworkspace-backup.adoc

:_content-type: CONCEPT
:description: DevWorkspace backup
:keywords: backup, storage
:navtitle: DevWorkspace backup
//:page-aliases:

[id="devworkspace-backup"]
= Workspace backup for {prod}

The {devworkspace} backup job allows for periodic backups of {devworkspace} data to a specified backup location.
Once enabled and configured, the backup job will run at defined intervals to create backups of {devworkspace} data.
The backup controller depends on an OCI-compatible registry e.g.,

> https://docs.redhat.com/en/documentation/openshift_container_platform/4.8/html/registry/registry-overview#registry-integrated-openshift-registry_registry-overview[OpenShift build-in registry]

Reviewer: ```suggestion
https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp4-ver}/html/registry/registry-overview#registry-integrated-openshift-registry_registry-overview[OpenShift build-in registry]
```

Author: Fixed

or https://quay.io[quay.io] used as an image artifact storage for backup archives.

The backup makes a snapshot of Workspace PVCs and stores them as tar.gz archives in the specified OCI registry.

> NOTE: By default, the {devworkspace} backup job is disabled.

Reviewer: ```suggestion

NOTE: By default, the {devworkspace} backup job is disabled.
```

Author: Fixed

The backup is configurable using `DevWorkspaceOperatorConfig` with the following fields:

* *`enable`*: Set to `true` to enable the backup job, `false` to disable it. Default: `false`.

> * *`schedule`*: A Cron expression defining how often the backup job runs. Default: `"0 1 * * *"`.

Reviewer 2: Not related to this PR but I noticed minor discrepancy when I was checking default value for `schedule`:

- [kubebuilder annotation](https://github.com/devfile/devworkspace-operator/blob/9b46583ecde32f8a73802487df1c31864d505108/apis/controller/v1alpha1/devworkspaceoperatorconfig_types.go#L70) in CRD definition sets default value as `0 1 * * *`
- [defaults.go](https://github.com/devfile/devworkspace-operator/blob/9b46583ecde32f8a73802487df1c31864d505108/pkg/config/defaults.go#L95) sets default value as `0 0 1 * *`

I think it's correctly set as kubebuilder annotation would be the one that would be used when CRD gets installed. However, do you think it's okay to have this difference in CRD default value and programmatic value?

Author: I am not sure what you mean here. When looking at the latest annotation the is `// +kubebuilder:default:="0 0 1 * *"` which matches the value from `defaults.go`

Reviewer 2: Sorry, I shared link for CleanupCronJob instead. Are you able to see `0 1 * * *` schedule in BackupCronJob config https://github.com/devfile/devworkspace-operator/blob/main/apis/controller/v1alpha1/devworkspaceoperatorconfig_types.go#L111 ?

Author: Fixed here: https://github.com/devfile/devworkspace-operator/pull/1602

* *`registry.path`*: A base registry location where the backup archives will be pushed.
+
The value provided for registry.path is only the first segment of the final location. The full registry path is assembled dynamically, incorporating the name of the workspace and the :latest tag, following this pattern:
+
`<registry.path>/<namespace>/<devworkspace-name>:latest`

* *`registry.authSecret`*: (Optional) The name of the Kubernetes Secret containing credentials to access the OCI registry. If not provided, it is assumed that the registry is public or uses integrated OpenShift registry.
* *`oras.extraArgs`*: (Optional) Additional arguments to pass to the `oras` CLI tool during push and pull operations.

There are several configuration options to customize the logic:

== Integrated OpenShift container registry

> This option is available only on OpenShift clusters with integrated container registry enabled and requires no additional configuration.

Reviewer: It would be nice to have a cross link to enabling OpenShift image registry

Author: Added

To enable the backup use following configuration in the global DWOC:

[source,yaml,subs="+attributes,+quotes"]
----
apiVersion: controller.devfile.io/v1alpha1
kind: DevWorkspaceOperatorConfig
metadata:
  name: devworkspace-operator-config
  namespace: $OPERATOR_INSTALL_NAMESPACE
config:
  routing:
    defaultRoutingClass: basic
  workspace:
    backupCronJob:
      enable: true
      registry:
        path: default-route-openshift-image-registry.apps.{cluster ID}.openshiftapps.com
      schedule: '0 */4 * * *' # cron expression with backup frequency
    imagePullPolicy: Always
----

> **Note:** The `path` field must contain the URL to your OpenShift integrated registry given by the cluster.

Reviewer: Should we mention how to get path?
```
echo "https://$(oc get route default-route -n openshift-image-registry --template='{{ .spec.host }}'))"
```

Author: Added

> 

Reviewer 3: Should we mention at this point that a backup job is triggered by stopping the workspace?

Author: I added a note in the generic backup section that only stopped workspaces are backed up.

Once the backup job is finished, the backup archives will be available in the {devworkspace} {namespace} under a repository
with a matching {devworkspace} name.

== Regular OCI-compatible registry
To use a regular OCI-compatible registry for backups, you need to provide registry credentials. Depending on your
RBAC policy, the token can be provided via a secret in the Operator {namespace} or in each {devworkspace} {namespace}.
Having the secret in the {devworkspace} {namespace} allows for using different registry accounts per {namespace} with more
granular access control.

[source,yaml,subs="+attributes,+quotes"]
----
kind: DevWorkspaceOperatorConfig
apiVersion: controller.devfile.io/v1alpha1
metadata:
  name: devworkspace-operator-config
  namespace: $OPERATOR_INSTALL_NAMESPACE
config:
  routing:
    defaultRoutingClass: basic
  workspace:
    backupCronJob:
      enable: true
      registry:
        authSecret: my-secret
        path: quay.io/my-company-org
      schedule: '0 */4 * * *'
    imagePullPolicy: Always
----
The `authSecret` must point to a real {kubernetes} Secret of type `kubernetes.io/dockerconfigjson` containing credentials to access the registry.

> To create one you can use following command:

Reviewer 2: ```suggestion
To create one, you can use the following command:
```

Author: Fixed

:_content-type: CONCEPT
:description: DevWorkspace backup
:keywords: backup, storage
:navtitle: DevWorkspace backup
//:page-aliases:

[id="devworkspace-backup"]
= Workspace backup for {prod}

> The {devworkspace} backup job allows for periodic backups of {devworkspace} data to a specified backup location.

Reviewer 4: ```suggestion
The {devworkspace} backup job provides periodic backups of {devworkspace} data to a specified location. After you enable and configure the job, it runs at defined intervals to create backups of {devworkspace} data. The backup controller requires an OCI-compliant registry, such as the **Red Hat OpenShift** integrated registry or link:https://quay.io[Quay.io], to store backup archives as image artifacts.
```

Author: Fixed

> Once enabled and configured, the backup job will run at defined intervals to create backups of {devworkspace} data.

Reviewer 4: ```suggestion

```

Author: Fixed

> The backup controller depends on an OCI-compatible registry e.g.,https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp4-ver}/html/registry/registry-overview#registry-integrated-openshift-registry_registry-overview[OpenShift build-in registry]

Reviewer 4: ```suggestion

```

Author: Fixed

> or https://quay.io[quay.io] used as an image artifact storage for backup archives.

Reviewer 4: ```suggestion

```

Author: Fixed

> The backup makes a snapshot of *stopped* Workspace PVCs and stores them as tar.gz archives in the specified OCI registry.

Reviewer 4: ```suggestion
The backup process takes a snapshot of *stopped* workspace persistent volume claims (PVCs) and stores them as `tar.gz` archives in the specified OCI registry.
```

Author: @gtrivedi88  The *stopped* workspace is impotant detail to mention. Do you really want it to be removed? The backup won't happen for the running workspace.

Reviewer 4: Thank you, please add. I have updated my comment.

Author: Updated

NOTE: By default, the {devworkspace} backup job is disabled.

> The backup is configurable using `DevWorkspaceOperatorConfig` with the following fields:

Reviewer 4: ```suggestion
You can configure the backup by using the `DevWorkspaceOperatorConfig` resource with the following fields:
```

Author: Fixed

> * *`enable`*: Set to `true` to enable the backup job, `false` to disable it. Default: `false`.

Reviewer 4: ```suggestion
* `enable`: Set to `true` to enable the backup job or `false` to disable it. The default value is `false`.
```

Author: Fixed

> * *`schedule`*: A Cron expression defining how often the backup job runs. Default: `"0 1 * * *"`.

Reviewer 4: ```suggestion
* `schedule`: A Cron expression that defines the backup frequency. The default value is `"0 1 * * *"`.
```

Author: Fixed

> * *`registry.path`*: A base registry location where the backup archives will be pushed.

Reviewer 4: ```suggestion
* `registry.path`: The base registry location for backup archives.
```

Author: Fixed

+

> The value provided for registry.path is only the first segment of the final location. The full registry path is assembled dynamically, incorporating the name of the workspace and the :latest tag, following this pattern:

Reviewer 4: ```suggestion
The value for `registry.path` is the first segment of the final location. The full path is assembled dynamically by using the workspace name and the `:latest` tag in the following pattern:
```

Author: Fixed

+
`<registry.path>/<namespace>/<devworkspace-name>:latest`

> * *`registry.authSecret`*: (Optional) The name of the Kubernetes Secret containing credentials to access the OCI registry. If not provided, it is assumed that the registry is public or uses integrated OpenShift registry.

Reviewer 4: ```suggestion
* `registry.authSecret`: (Optional) The name of the Kubernetes secret that contains credentials for the OCI registry. If you do not provide a secret, the system assumes the registry is public or uses the **Red Hat OpenShift** integrated registry.
```

Author: Fixed

> * *`oras.extraArgs`*: (Optional) Additional arguments to pass to the `oras` CLI tool during push and pull operations.

Reviewer 4: ```suggestion
* `oras.extraArgs`: (Optional) Additional arguments for the `oras` CLI tool during push and pull operations.
```

Author: Fixed

> There are several configuration options to customize the logic:

Reviewer 4: I feel this is a big topic and we can have multiple .adoc files.
Please see my comment here: https://github.com/eclipse-che/che-docs/pull/3029#discussion_r2848863817

I see we have a mix of concept and procedures in one .adoc file.

Author: I split the doc into multiple as suggested.

:_content-type: CONCEPT
:description: DevWorkspace backup
:keywords: backup, storage
:navtitle: DevWorkspace backup
//:page-aliases:

[id="devworkspace-backup"]
= Workspace backup for {prod}

> Starting from {prod} version v7.115.0, the {devworkspace} backup job allows for periodic backups of {devworkspace} data to a specified backup location.

Reviewer 5: @gtrivedi88 ptal, and recommend how the docs should be downstreamed e.g. 7.115 - > Dev Spaces 3.27

Reviewer 4: @ibuziuk As for downstreaming, when this is merged and we carve out a 7.115 branch in main. The [pipeline watcher](https://gitlab.cee.redhat.com/red-hat-developers-documentation/devspaces-documentation/-/pipeline_schedules) which watches for the changes in upstream branch for specific release versions, will [create a issue](https://gitlab.cee.redhat.com/red-hat-developers-documentation/devspaces-documentation/-/issues) in Gitlab.

At the very begining of every release, in the [attributes](https://gitlab.cee.redhat.com/red-hat-developers-documentation/devspaces-documentation/-/blob/main/common/attributes.adoc?ref_type=heads&plain=1)  we have to define the upstream branches (7.114.x, and 7.115.x) which will make downstream (3.27) and then schedule a pipieline, which will keep watching for those branches for content updates.

Let me know if you have questions

Once enabled and configured, the backup job will run at defined intervals to create backups of {devworkspace} data.

> The backup controller depends on an OCI-compatible registry e.g., https://quay.io[quay.io] used as an image artifact storage for backup archives.

Reviewer 5: should we mention OpenShift registry at the first place

Author: Added
